server/src/main/java/io/atomix/copycat/server/state/ConnectionManager.java

```diff
@@ -31,6 +31,7 @@
 public final class ConnectionManager {
   private final Client client;
   private final Map<Address, Connection> connections = new HashMap<>();
+  private final Map<Address, CompletableFuture<Connection>> futures = new HashMap<>();
```

Comment:
Just my personal preference, so take it or leave it:
Maybe change futures to something more descriptive like "connectionFutures"?